In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [4]:
df = pd.read_csv("../datasets/diabetes.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [6]:
X = df.iloc[:, :8]
X.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [10]:
y = df.Outcome
y.head()
print(f'X Shape: {X.shape} \nY Shape: {y.shape}')

X Shape: (768, 8) 
Y Shape: (768,)


In [11]:
scaler = StandardScaler()
X_std = scaler.fit_transform(X)

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X_std, y, test_size=0.3, random_state=0)

In [15]:
print(f'X_train Shape: {X_train.shape} \ny_train Shape: {y_train.shape}')
print(f'X_test Shape: {X_test.shape} \ny_test Shape: {y_test.shape}')

X_train Shape: (537, 8) 
y_train Shape: (537,)
X_test Shape: (231, 8) 
y_test Shape: (231,)


In [52]:
mlp = MLPClassifier(hidden_layer_sizes=(55), activation='relu', solver='sgd', max_iter=200, verbose=True)
mlp

MLPClassifier(hidden_layer_sizes=55, solver='sgd', verbose=True)

In [53]:
mlp.fit(X_train, y_train)

Iteration 1, loss = 0.72055495
Iteration 2, loss = 0.71761277
Iteration 3, loss = 0.71303016
Iteration 4, loss = 0.70749152
Iteration 5, loss = 0.70136113
Iteration 6, loss = 0.69458552
Iteration 7, loss = 0.68798999
Iteration 8, loss = 0.68094056
Iteration 9, loss = 0.67447294
Iteration 10, loss = 0.66800426
Iteration 11, loss = 0.66169707
Iteration 12, loss = 0.65525938
Iteration 13, loss = 0.64940888
Iteration 14, loss = 0.64366991
Iteration 15, loss = 0.63818390
Iteration 16, loss = 0.63295990
Iteration 17, loss = 0.62813833
Iteration 18, loss = 0.62317806
Iteration 19, loss = 0.61856772
Iteration 20, loss = 0.61428557
Iteration 21, loss = 0.61003469
Iteration 22, loss = 0.60610941
Iteration 23, loss = 0.60244527
Iteration 24, loss = 0.59859903
Iteration 25, loss = 0.59516379
Iteration 26, loss = 0.59179953
Iteration 27, loss = 0.58862057
Iteration 28, loss = 0.58546287
Iteration 29, loss = 0.58252634
Iteration 30, loss = 0.57964720
Iteration 31, loss = 0.57699860
Iteration 32, los

C:\Users\adity\anaconda3\envs\ml\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(hidden_layer_sizes=55, solver='sgd', verbose=True)

In [54]:
print(mlp.score(X_train,y_train)*100)

77.46741154562383


In [55]:
print(mlp.score(X_test, y_test)*100) # Overfitting

77.05627705627705


In [56]:
y_pred = mlp.predict(X_test)
y_pred

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0], dtype=int64)

In [57]:
y_test.ravel()

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0], dtype=int64)

In [60]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[140  17]
 [ 36  38]]


### Confusion Matrix
| Not Diabetic  -  140 | Not Diabetic but Shows you have Diabetes  -  17
| ----------- | ----------- |
| Diabetic but shows not Diabetic  -  36 | Diabetic  -  38 |

In [59]:
cr = classification_report(y_test, y_pred)
print(cr)

              precision    recall  f1-score   support

           0       0.80      0.89      0.84       157
           1       0.69      0.51      0.59        74

    accuracy                           0.77       231
   macro avg       0.74      0.70      0.71       231
weighted avg       0.76      0.77      0.76       231



In [62]:
X_test

array([[-0.84488505,  2.44447821,  0.35643175, ...,  1.38436175,
         2.784923  , -0.95646168],
       [-0.54791859, -0.43485916,  0.25303625, ...,  0.20401277,
        -0.20499449, -0.87137393],
       [ 0.04601433, -1.40507067, -0.36733675, ...,  0.25478047,
        -0.24425603, -0.70119842],
       ...,
       [-0.84488505, -0.81042491, -0.47073225, ..., -1.02710391,
        -0.63989158, -0.95646168],
       [-0.84488505, -0.46615631,  0.35643175, ...,  0.69899783,
        -0.83015905, -0.61611067],
       [ 1.82781311,  0.03459802,  0.45982725, ..., -0.5575027 ,
         0.12117833,  1.00055664]])

In [63]:
df.tail()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

In [66]:
sample = np.array([5,121,72,23,	112,26.2,0.245,30])
sample.shape

(8,)

In [68]:
print(f'X Test Shape: {X_test.shape}')

X Test Shape: (231, 8)


In [69]:
sample = np.expand_dims(sample, 0)
sample.shape

(1, 8)

In [71]:
print(mlp.predict(sample)) # wrong prediction

[1]


In [75]:
sample2 = np.array([5,126,60,0,0,30.1,0.349,47])
sample2 = np.expand_dims(sample2, 0)
print(mlp.predict(sample2))

[1]
